tools/driver/p4c/main.py

```diff
@@ -0,0 +1,190 @@
+    args = opts.source_file
+
+    if opts.show_version:
+        print("p4c %s" % (p4c.__version__))
+        sys.exit(0)
+
+    if not args:
+        parser.error('No input specified.')
+
+    init_globals()
+
+    save_options("preprocessor", opts.preprocessor_options)
+    save_options("compiler", opts.compiler_options)
+    save_options("assembler", opts.assembler_options)
+    save_options("linker", opts.linker_options)
+
+    # set output directory
+    if not os.path.exists(opts.output_directory):
+        os.makedirs(opts.output_directory)
+
+    if "P4C_INCLUDE_PATH" not in os.environ:
+        print "error: P4C_INCLUDE_PATH undefined."
+        sys.exit(1)
+
+    cfg_files = glob.glob("{}/*.cfg".format(os.path.dirname(__file__)))
+    cfg = p4c.config.Config(config_prefix = "p4c")
+    for cf in cfg_files:
+        cfg.load_from_config(cf, opts.output_directory, opts.source_file)
+
+
+    if opts.backend not in cfg.steps:
+        print "Unknown target:", opts.backend
+        sys.exit(1)
+
+    commands['preprocessor'].insert(0, cfg.get_preprocessor(opts.backend))
+    commands['compiler'].insert(0, cfg.get_compiler(opts.backend))
+    commands['assembler'].insert(0, cfg.get_assembler(opts.backend))
+    commands['linker'].insert(0, cfg.get_linker(opts.backend))
+
+    # handle mode flags
+    step_enable = [False, False, False, False]
+    if opts.run_preprocessor_only:
+        step_enable = [True, False, False, False]
+    elif opts.skip_preprocessor:
+        step_enable = [False, True, True, True]
+    elif opts.run_till_assembler:
+        step_enable = [True, True, False, False]
+    elif opts.run_all:
+        step_enable = [True, True, True, True]
+
+    # default search path
+    for path in os.environ['P4C_INCLUDE_PATH'].split(':'):
+        if (opts.language == 'p4-16' and os.path.isdir(path + '/p4include')):
+            commands['preprocessor'].append("-I {}".format(path+'/p4include'))
+            commands['compiler'].append("-I {}".format(path+'/p4include'))
+        elif (opts.language == 'p4-14' and os.path.isdir(path + 'p4_14include')):
+            commands['preprocessor'].append("-I {}".format(path+'/p4_14include'))
+            commands['compiler'].append("-I {}".format(path+'/p4_14include'))
+
+    # append search path
+    for path in opts.search_path:
+        commands['preprocessor'].append("-I")
+        commands['preprocessor'].append(path)
+        commands['compiler'].append("-I")
+        commands['compiler'].append(path)
+
+    # p4 version
+    if opts.language == 'p4-16':
+        commands['compiler'].append("--p4v=16")
+    else:
+        commands['compiler'].append("--p4v=14")
+
+    for idx, step in enumerate(cfg.steps[opts.backend]):
+        cmd = commands[step]
+        options = cfg.options[opts.backend][step]
+        cmd = cmd + options
+        # convert ['-a -b', '-c', '-d'] to ['-a', '-b', '-c', '-d']
+        cmd = list(itertools.chain.from_iterable(el.split() for el in cmd))
```

Comment:
This doesn't handle quotes right. =) Try this:

https://docs.python.org/2/library/shlex.html